File: Core/src/com/stusherwin/midiapp/core/InputNode.java
package com.stusherwin.midiapp.core;

import rx.Observable;
import rx.schedulers.Schedulers;

import java.util.ArrayList;
import java.util.List;

public class InputNode<T> extends Node<T> {
    private List<OutputNode> _connected = new ArrayList<OutputNode>();

    public void connectTo(OutputNode<T> output) {
        if( !_connected.contains(output)) {
            _connected.add(output);
            output.subject
                    .subscribeOn(Schedulers.newThread())
                    .subscribe(this.subject);
        }
    }

    public Observable<T> observable() {
        return this.subject;
    }
}

File: Core/src/com/stusherwin/midiapp/core/modules/InputNode.scala
package com.stusherwin.midiapp.core.modules

import rx.lang.scala.Observable
import rx.lang.scala.schedulers.NewThreadScheduler

class InputNode[T] extends Node[T] {
  def observable : Observable[T] = this.subject

  def connectTo(output : OutputNode[T]) : Unit = {
    output.subject
      .subscribeOn(NewThreadScheduler.apply())
      .subscribe(this.subject)
  }
}